Source/EasyNetQ/Producer/PublishExchangeDeclareStrategy.cs

```diff
@@ -7,24 +8,56 @@ namespace EasyNetQ.Producer
 {
     public class PublishExchangeDeclareStrategy : IPublishExchangeDeclareStrategy
     {
-        private readonly ConcurrentDictionary<string, Task<IExchange>> exchangeNames =new ConcurrentDictionary<string, Task<IExchange>>();
-     
+        private readonly ConcurrentDictionary<string, IExchange> exchanges = new ConcurrentDictionary<string, IExchange>();
+        private readonly SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
+
         public IExchange DeclareExchange(IAdvancedBus advancedBus, string exchangeName, string exchangeType)
         {
-            return DeclareExchangeAsync(advancedBus, exchangeName, exchangeType).Result;
+            IExchange exchange;
+            if(exchanges.TryGetValue(exchangeName, out exchange))
+                return exchange;
+            semaphore.Wait();
+            try
+            {
+                if (exchanges.TryGetValue(exchangeName, out exchange))
+                    return exchange;
+                exchange = advancedBus.ExchangeDeclare(exchangeName, exchangeType);
+                exchanges[exchangeName] = exchange;
+                return exchange;
+            }
+            finally
+            {
+                semaphore.Release();
+            }
         }
         
         public IExchange DeclareExchange(IAdvancedBus advancedBus, Type messageType, string exchangeType)
         {
-            return DeclareExchangeAsync(advancedBus, messageType, exchangeType).Result;
+            var conventions = advancedBus.Container.Resolve<IConventions>();
+            var exchangeName = conventions.ExchangeNamingConvention(messageType);
+            return DeclareExchange(advancedBus, exchangeName, exchangeType);
         }
 
-        public Task<IExchange> DeclareExchangeAsync(IAdvancedBus advancedBus, string exchangeName, string exchangeType)
+        public async Task<IExchange> DeclareExchangeAsync(IAdvancedBus advancedBus, string exchangeName, string exchangeType)
         {
-            return exchangeNames.AddOrUpdate(
-                exchangeName,
-                name => advancedBus.ExchangeDeclareAsync(name, exchangeType),
-                (name, exchangeTask) => exchangeTask.IsFaulted ? advancedBus.ExchangeDeclareAsync(name, exchangeType) : exchangeTask);
+            IExchange exchange;
+            if (exchanges.TryGetValue(exchangeName, out exchange))
+                return exchange;
+            // TODO Need async waiting here
+            // http://blogs.msdn.com/b/pfxteam/archive/2012/02/12/10266983.aspx
```

Comment:
@micdenny yeap, will send a patch today :)
Also there is one more thing to do: queue.Add in client command dispatcher blocks the calling thread, we also can try to use AsyncQueue.